File: DSA/Assignment 4/2.c
#include "singlylinkedlist.h"
NODE *sortedMerge(NODE *, NODE *);
int main()
{
    NODE *head1 = NULL, *head2 = NULL;
    printf("Enter First List\n");
    head1 = createList(head1);
    displayList(head1);
    printf("\nEnter Second List\n");
    head2 = createList(head2);
    displayList(head2);
    NODE *res = sortedMerge(head1, head2);
    printf("\nSorted Merge list is\n");
    displayList(res);

    return 0;
}
NODE *sortedMerge(NODE *list1, NODE *list2)
{
    if (list1 == NULL)
        return list2;
    if (list2 == NULL)
        return list1;
    if (list1->data > list2->data)
    {
        NODE *temp = list1;
        list1 = list2;
        list2 = temp;
    }
    NODE *res = list1;
    while (list1 != NULL && list2 != NULL)
    {
        NODE *tmp = NULL;
        while (list1 != NULL && list1->data <= list2->data)
        {
            tmp = list1;
            list1 = list1->next;
        }
        tmp->next = list2;
        // swap
        NODE *temp = list1;
        list1 = list2;
        list2 = temp;
    }
    return res;
}

File: DSA/Assignment 4/python/2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def createList(self):
        print("Creating Linked List...")
        print("Enter -1 to end")
        num = int(input("Enter the data: "))
        while(num != -1):
            newNode = Node(num)
            if self.head is None:
                self.head = newNode
            else:
                last = self.head
                while last.next:
                    last = last.next
                last.next = newNode
            num = int(input("Enter the data: "))

    def displayList(self):
        if self.head is None:
            print("List is empty")
            return
        print("HEAD", end='')
        temp = self.head
        while temp:
            print(" -> " + str(temp.data), end='')
            temp = temp.next
        print(" -> NULL\n")


def mergeLists(headA, headB):
    dummyNode = Node(0)
    tail = dummyNode
    while True:
        if headA is None:
            tail.next = headB
            break
        if headB is None:
            tail.next = headA
            break
        if headA.data <= headB.data:
            tail.next = headA
            headA = headA.next
        else:
            tail.next = headB
            headB = headB.next
        tail = tail.next
    return dummyNode.next


listA = LinkedList()
print("Enter First List")
listA.createList()
listA.displayList()

listB = LinkedList()
print("Enter Second List")
listB.createList()
listB.displayList()

listA.head = mergeLists(listA.head, listB.head)
print("Merged Linked List is:")
listA.displayList()